Feature/ allow ulimits parameters configuration of the devices

g4br-i opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/KatharaFramework/Kathara/blob/master/CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

I've introduced the possibility to set ulimits in the devices via lab.conf file

**-Changed files**
The following files have been modified:
- src/Kathara/manager/docker/DockerMachine.py:
  - to be able to support the ulimits instances upon container creation.
  - _create_ulimit_instances method was introduced to parse a ulimits dictionary to ulimit instances following the docker sdk requirements
- src/Kathara/model/Machine.py:
  - the add_meta method is now able to parse strings  like _yourdevice_[ulimit]="_ulimit_=_softlimit_:_hardlimit_" or _yourdevice_[ulimit]="_ulimit_=_softlimit_" in the second case _hardlimit_ will be set to be equal to _softlimit_
  - get_ulimits method was introduced to retrieve the dictionary of ulimits
- tests/model/machine_test.py, added tests to verify correct parsing 
- docs/kathara-lab.conf.5.ronn, added some lines to explain how to set ulimits

**- How to verify it**
- Create a lab.conf with a some devices then set _yourdevice_[ulimit]="nofile=1000:1000" or _yourdevice_[ulimit]="memlock=-1"
- start the lab
- in the terminal of the device run ulimit -a 
- check that ulimit you've set in lab conf matches printed value in the teminal

**- Description for the changelog**
Lab.conf now supports ulimits configuration at device creation, this solves cases where specific containers need to be run with custom values as in "docker run --ulimit memlock=819200000:819200000 ..."

g4br-i committed:
adding ulimit to .conf file so devices can modify ulimits

g4br-i committed:
tested

g4br-i committed:
if soft limit is bigger than hard limit, soft limit will be set to be equal to  hard limit

Skazza94 commented on tests/model/machine_test.py:
```diff
@@ -282,6 +283,26 @@ def test_add_meta_overwrite_port(default_device: Machine):
+def test_add_meta_ulimit_not_format_exception(default_device):
+    with pytest.raises(MachineOptionError):
+        default_device.add_meta("ulimit", "nofile1024")
+
```
Can you add a test with `soft > hard`? (it's the only case not tested)

g4br-i replied:
while adding that test i realized i was not covering some edge cases like people giving less than -1 or cases where hard = -1 and limit is a number different from -1 which was not properly handled . I think now i've addressed all cases and also properly added tests.

Skazza94 reviewed (changes requested):
Hi @g4br-i,
thanks for the fantastic PR. Everything looks perfect! :smile: 

I have just one minor comment on the tests.

Also, the `ulimits` meta is not compatible with Kubernetes. Perhaps, maybe it is better to include a warning to notify the user that the meta will be ignored (like the `bridged` option).  Can you include a warning in `src/Kathara/manager/kubernetes/KubernetesMachine.py` similar to this one? 
https://github.com/KatharaFramework/Kathara/blob/8f63323ca864c71bdb7191af7a12166f250128b0/src/Kathara/manager/kubernetes/KubernetesMachine.py#L304

After line 304 it should be fine.

Thanks!

g4br-i committed:
addressing other possible error cases when setting ulimits, added a warning for kubernets, added multiple test cases

g4br-i commented:
with the latest commit i've added:

[https://github.com/KatharaFramework/Kathara/blob/c9f7d13a5a1ec7a4e12a84ee07841488d9b4b3cf/src/Kathara/model/Machine.py#L212-L223](url)
this addresses some edge case that i did not consider initially, like setting ulimit less the -1 and fixing a wrong behaviour i've introduced by imposing only that hard must be bigger than soft ( i was not considering the case hard=-1 which is equal to unlimited and soft being a value like 1024).

i've also added the warning for kubernets

[https://github.com/KatharaFramework/Kathara/blob/c9f7d13a5a1ec7a4e12a84ee07841488d9b4b3cf/src/Kathara/manager/kubernetes/KubernetesMachine.py#L306-L308](url)

also a bunch of new tests:

[https://github.com/KatharaFramework/Kathara/blob/c9f7d13a5a1ec7a4e12a84ee07841488d9b4b3cf/tests/model/machine_test.py#L306C1-L324C60](url)

tcaiazzi reviewed (approved):
Super thanks for the PR!

Skazza94 reviewed (approved):
Hi @g4br-i,
thanks for the fixes!

The PR is good to go. I will merge it in the `develop` branch and the feature will be available in the next Kathará release.
